Draft coding and design guidelines (#2580)
Particular/ServiceControl: docs/coding-and-design-guidelines.md

- **Prefer Hosted Services over NServiceBus FeatureStartupTasks**: ServiceControl includes a lot of background tasks and these are both abstractions for building background tasks. `FeatureStartupTask` implementations are tied to the lifecycle of an endpoint. In general, we prefer to use the `IHostedService` abstraction which is tied to the lifecycle of the host application. NOTE: Hosted Services are started in the order that they are registered which allows us to better control the startup sequence. They are shut down in the reverse order.
- **Prefer IServiceCollection over IConfigureComponents or IContainerBuilder**: Where possible we'd rather use the generic Microsoft DI abstractions over the NServiceBus one (IConfigureComponents) or the Autofac one (IContainerBuilder).

>   - When registering components from within an NServiceBus Feature, it still  makes sense to use IConfigureComponents but we should consider whether it makes sense to move the code out of NServiceBus features

Reviewer: Does it make sense to add a note here to mention that this will be “solved” when SC will be migrated to v8, since v8 uses IServiceCollection by default?

  - When relying on Autofac specific features, it makes sense to rely on IContainerBuilder. If a feature can be implemented in a way to avoid this, it should be preferred. We may in the future decide to remove our dependency on Autofac but that decision has not been made.

> - **Prefer IServiceProvider over ILifetimeService over IContainer**: As above, we'd rather rely on the generic Microsoft DI abstraction where possible and only fall back to NServiceBus or Autofac abstractions where strictly necessary.

Reviewer: Same as above.

> - **Prefer IServiceProvider over ILifetimeService over IContainer**: As above, we'd rather rely on the generic Microsoft DI abstraction where possible and only fall back to NServiceBus or Autofac abstractions where strictly necessary.

Reviewer 2: ```suggestion
- **Prefer `IServiceProvider` over `ILifetimeService` and `IContainer`**: As above, we prefer to use the Microsoft DI abstraction where possible and only fall back `ILifetimeService` where strictly necessary.
```

>   - If relying on a specific feature of Autofac, then `ILifetimeService` should be used.

Reviewer 2: ```suggestion
  - When relying on a Autofac-specific feature, `ILifetimeService` should be used.
```

  - `IContainer` should never be used. It is functionally equivalent to `IServiceProvider`

### Exceptions

> There is one exception to the preference for Microsoft abstractions

Reviewer 3: This doesn't sound like an exception to the preference, but rather a practical "it will take time to migrate to the preferred abstraction"

Reviewer 4: Agreed, perhaps say "until NServiceBus core removes its static logging"?

Side note: this is tracked here https://github.com/Particular/PlatformArchitecture/blob/master/problems/nservicebus-core.md#custom-logging-abstraction

Author: We didn't want to make the decision that we are definitely moving away from NServiceBus logging, especially as it has not been deprecated for NServiceBus version 8. With that in mind, it is an exception to the preference. In this one case we do not prefer the Microsoft abstration.

@WilliamBZA @andreasohlund Do you think we need to make a stronger statement about it now?

Reviewer 3: I'm not terribly concerned about it. If we haven't made a clear decision then that's fine too.

- **Use NServiceBus Logging abstractions over Microsoft or NLog abstractions**: The existing ServiceControl code uses the static `LogManager` classes to get access to the logging infrastructure and all new code should follow this pattern. In the future we are likely to switch to the Microsoft abstraction but until then we want to maintain consistency.

## Prefer explicit container registration

> Where possible, we prefer explicit container registration for services as opposed to convention-bassed registration. This allows us to better see which classes belong to which ServiceControl components or which part of the ServiceControl infrastructure. It also allows us more explicit control over what services are available within the container which helps to reduce inappropriate cross-component access. In the future we may be able to make this more explicit by moving ServiceControl components into their own assemblies and keeping non-shared services internal.

Reviewer 3: ```suggestion
Where possible, we prefer explicit container registration for services as opposed to convention-based registration. This allows us to better see which classes belong to which ServiceControl components or which part of the ServiceControl infrastructure. It also allows us more explicit control over what services are available within the container which helps to reduce inappropriate cross-component access. In the future we may be able to make this more explicit by moving ServiceControl components into their own assemblies and keeping non-shared services internal.
```

> Where possible, we prefer explicit container registration for services as opposed to convention-bassed registration. This allows us to better see which classes belong to which ServiceControl components or which part of the ServiceControl infrastructure. It also allows us more explicit control over what services are available within the container which helps to reduce inappropriate cross-component access. In the future we may be able to make this more explicit by moving ServiceControl components into their own assemblies and keeping non-shared services internal.

Reviewer 2: ```suggestion
Where possible, we prefer explicit container registration for services instead of convention-based registration. This provides better visibility of which classes belong to which ServiceControl components or to which part of the ServiceControl infrastructure. It also gives us more explicit control over which services are available within the container, which helps to reduce inappropriate cross-component access. In the future we may be able to make this more explicit by moving ServiceControl components into their own assemblies and keeping non-shared services internal.
```

### Exceptions

> There are a few things that are still registered via convention. Note that these are all registered via Autofac and not via the Microsoft DI abstractions.

Reviewer 2: ```suggestion
There are a few things that are still registered using convention. Note that these are all registered using Autofac and not the Microsoft DI abstractions.
```

- **API Controllers**
- **Scatter-Gather API components**

Additionally, because NServiceBus does a type-scan at startup it will automatically register any implementations of `Feature` and `IHandleMessage<>`. We have chosen to leave this alone as we would be fighting with NServiceBus in order to turn this off. 

## Avoid property injection

> Although the Autofac container can be configured to allow property injection, we prefer to avoid it. There is no way to specify property injection using the generic microsoft DI abstractions, nor does the default Service Provider support it. Where possible, use constructor injection instead.

Reviewer 2: ```suggestion
Although the Autofac container can be configured to allow property injection, we prefer to avoid it. There is no way to specify property injection using the Microsoft DI abstractions, and the default `IServiceProvider` implementation does not support it. Where possible, use constructor injection instead.
```

### Exceptions

> There are a few places where property injection is still used. Note that these are all registered via Autofac and not via the Microsoft DI abstractions.

Reviewer 2: ```suggestion
There are a few places where property injection is still used. These are all registered using Autofac, and not the Microsoft DI abstractions.
```
